### DISCUSSION TITLE
Client Controlled Nullability: List Syntax

### DISCUSSION THREAD

**Discussion by twof** [graphql/graphql-wg#864] [General]
[Spec changes](https://github.com/graphql/graphql-spec/pull/895)
To get a better idea of where we stand, I'm including a poll so people can vote on their favorite option. This vote is non-binding, and is only one factor in choosing which syntax we choose I'll open new threads for people to discuss each option below.

`[[!]!]?`
[![](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/%60%5B%5B!%5D!%5D%3F%60)](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/%60%5B%5B!%5D!%5D%3F%60/vote)
(apply from the inside-out) `!!?`
[![](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/(apply%20from%20the%20inside-out)%20%60!!%3F%60)](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/(apply%20from%20the%20inside-out)%20%60!!%3F%60/vote)
(apply from the outside-in) `?!!`
[![](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/(apply%20from%20the%20outside-in)%20%60%3F!!%60)](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/(apply%20from%20the%20outside-in)%20%60%3F!!%60/vote)
`<= [[Int!]!]`/`as [[Int!]!]`
[![](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/%3C%3D%20%5B%5BInt!%5D!%5D)](https://api.gh-polls.com/poll/01FSQB7G1G5461RAPZVMEN16B7/%3C%3D%20%5B%5BInt!%5D!%5D/vote)

Update (March 6th, 2022): Square bracket syntax has been selected for development, likely with some alterations made to accommodate the changes to null propagation behavior that was also selected at the last working group meeting.

**Comment by twof** [comment]
## `[[!]!]?`

**Reply by twof** [reply]
Previous discussions:
https://github.com/graphql/graphql-spec/pull/895#issuecomment-961442966
https://github.com/graphql/graphql-spec/pull/895#issuecomment-961461401
https://github.com/graphql/graphql-spec/pull/895#issuecomment-964741557
https://github.com/graphql/graphql-spec/pull/895#issuecomment-969921630
https://github.com/graphql/graphql-spec/pull/895#issuecomment-985357972

**Reply by leebyron** [reply]
For posterity since we got to discuss live:

This is the variant of those proposed here I think is most compelling, but I would suggest we alter its current form slightly.

Specifically, I don't think we should *require* brackets for fields with list types. It's reasonable that you may want to only apply CCN to the outer most part of the type and leave the inner part alone.

Also, especially since the specific behavior that `!` and `?` imply are still getting worked out, it might not be right to require they be written in since that may force users to apply a behavior that they do not want

**Reply by twof** [reply]
Sounds good! I'll work on loosening this requirement once I'm done with the null propagation change.

**Comment by twof** [comment]
## `(apply from the inside-out) !!?`/ `(apply from the outside-in) ?!!`

**Reply by twof** [reply]
Previous discussions:
https://github.com/graphql/graphql-spec/pull/895#issuecomment-961442966
https://github.com/graphql/graphql-spec/pull/895#issuecomment-985357972
https://github.com/graphql/graphql-spec/pull/895#issuecomment-985976949

**Reply by leebyron** [reply]
The fact that there's two ways to interpret this is probably a good sign that this isn't the right path

**Comment by twof** [comment]
## `<= [[Int!]!]`/`as [[Int!]!]`

**Reply by twof** [reply]
Previous discussions:
https://github.com/graphql/graphql-spec/pull/895#issuecomment-961461401
https://github.com/graphql/graphql-spec/pull/895#issuecomment-987899922
https://github.com/graphql/graphql-spec/pull/895#issuecomment-988153535
https://github.com/graphql/graphql-spec/pull/895#issuecomment-988179604
https://github.com/graphql/graphql-spec/pull/895#issuecomment-998555982
https://github.com/graphql/graphql-spec/pull/895#issuecomment-998940369

**Reply by leebyron** [reply]
The repeating types here read like a type cast instead of a null operator, and pretty different from precedent from other languages. While interesting, I think that opens a can of worms

**Reply by benjie** [reply]
It does, however, preserve option value for that :wink:
